pkg/datapath/loader/netlink.go

```diff
@@ -98,25 +96,36 @@ func replaceDatapath(ctx context.Context, ifName, objPath, progSec, progDirectio
 		}
 	}
 
+	// If the iproute2 call below is successful, any 'pending' map pins will be removed.
+	// If not, any pending maps will be re-pinned back to their initial paths.
+	var revert bool
 	cmd := exec.CommandContext(ctx, loaderProg, args...).WithFilters(libbpfFixupMsg)
 	if _, err := cmd.CombinedOutput(log, true); err != nil {
 		// Program/object replacement unsuccessful, revert bpffs migration.
 		revert = true
-		return fmt.Errorf("Failed to load prog with %s: %w", loaderProg, err)
+		return nil, fmt.Errorf("Failed to load prog with %s: %w", loaderProg, err)
 	}
 
-	return nil
+	finalize := func() {
+		l := log.WithField("device", ifName).WithField("objPath", objPath)
+		l.Debug("Finalizing bpffs map migration")
+		if err := bpf.FinalizeBPFFSMigration(bpf.MapPrefixPath(), objPath, revert); err != nil {
+			l.WithError(err).Error("Could not finalize bpffs map migration")
+		}
+	}
+
+	return finalize, nil
```

Comment:
The comment on line 99 states that if the load fails, the original maps will be repinned to their original paths. However, in that case (lines 104-106) the code only sets `revert = true` and then returns a `nil` finalizer. As a result, `bpf.FinalizeBPFFSMigration()` does not get called.

In the successful case, I think that we need to defer the map migration finalization until after all of the programs have been loaded, so that case makes sense to pass the finalizer back to the caller.

In the error case where the program cannot be loaded, I think it is OK to just run the contents of this finalizer function directly from here with `revert=true` and pass back a nil finalizer (like the code on line 106 does already).

Reply:
Oops, good catch. I opted to get rid of the bool and call finalize directly when handling the error. I also specified this behaviour in the `replaceDatapath` docstring.